File: scenarios/scenario-to-desktop/templates/build-tools/native-extension.test.ts
import * as fs from 'fs/promises';
import * as path from 'path';
import * as os from 'os';
import ts from 'typescript';
import { NativeExtensionCompatibilityError, validateNativeExtension } from './native-extension';
import { DesktopTemplateGenerator, DesktopConfig } from './template-generator';

const valid = { version: 1, module: 'presentation', permissions: ['window.presentation'], platforms: ['linux', 'win', 'mac'] };
describe('native extension admission', () => {
    it('requires version 3 and explicit context permission for native capture', () => {
        const extension = { ...valid, version: 3, permissions: ['window.presentation', 'global-shortcut', 'desktop.context'], activation_shortcut: 'Control+Shift+Space' };
        expect(() => validateNativeExtension(extension, 'electron', ['linux'])).not.toThrow();
        expect(() => validateNativeExtension({...extension, version: 2}, 'electron', ['linux'])).toThrow();
        expect(() => validateNativeExtension({...extension, permissions: ['window.presentation', 'global-shortcut']}, 'electron', ['linux'])).toThrow();
    });
    it('requires version 2 and its permission for activation', () => {
        const extension = { ...valid, version: 2, permissions: ['window.presentation','global-shortcut'], activation_shortcut: 'CommandOrControl+Shift+Space' };
        expect(() => validateNativeExtension(extension,'electron',['linux'])).not.toThrow();
        for (const invalid of [{...extension,version:1}, {...extension,permissions:['window.presentation']}, {...extension,activation_shortcut:'Alt+Alt+Space'}, {...extension,activation_shortcut:'code.js'}]) {
            expect(() => validateNativeExtension(invalid,'electron',['linux'])).toThrow();
        }
    });
    it('accepts canonical pipeline targets and refuses unknown architectures', () => {
        expect(() => validateNativeExtension(valid, 'electron', ['linux-amd64','linux-arm64','darwin-amd64','darwin-arm64','windows-amd64','windows-arm64'])).not.toThrow();
        expect(() => validateNativeExtension(valid, 'electron', ['linux-unknown'])).toThrow();
    });
    it('keeps vanilla configurations and admits the built-in module', () => {
        expect(() => validateNativeExtension(undefined, 'electron', ['linux'])).not.toThrow();
        expect(() => validateNativeExtension(valid, 'electron', ['linux', 'mac'])).not.toThrow();
    });
    it('returns a stable typed reason for an unsupported extension contract', () => {
        try {
            validateNativeExtension({ ...valid, version: 4 }, 'electron', ['linux']);
            throw new Error('expected extension admission to fail');
        } catch (error) {
            expect(error).toBeInstanceOf(NativeExtensionCompatibilityError);
            expect((error as NativeExtensionCompatibilityError).code).toBe('NATIVE_EXTENSION_UNSUPPORTED_CONTRACT');
        }
    });
    it('identifies a target mismatch without exposing arbitrary paths', () => {
        const linuxOnly = { ...valid, platforms: ['linux'] as ('linux' | 'mac' | 'win')[] };
        expect(() => validateNativeExtension(linuxOnly, 'electron', ['windows-arm64'])).toThrowError(NativeExtensionCompatibilityError);
        try {
            validateNativeExtension(linuxOnly, 'electron', ['windows-arm64']);
        } catch (error) {
            expect((error as NativeExtensionCompatibilityError).code).toBe('NATIVE_EXTENSION_TARGET_UNSUPPORTED');
            expect((error as NativeExtensionCompatibilityError).target).toBe('windows-arm64');
        }
    });
    it.each([
        { ...valid, version: 3 }, { ...valid, module: '../custom' },
        { ...valid, entrypoint: '/tmp/custom.js' }, { ...valid, permissions: ['filesystem'] },
        { ...valid, platforms: ['mac'] }, { ...valid, platforms: ['linux', 'linux'] },
    ])('rejects unsupported contracts before generation: %j', native_extension => {
        expect(() => new DesktopTemplateGenerator({ framework: 'electron', platforms: ['linux'], native_extension } as unknown as DesktopConfig)).toThrow();
    });
    it('admits only the governed desktop-session helper provider', () => {
        expect(() => validateNativeExtension({ ...valid, helper_providers: [{ owner: 'device-control', capability: 'desktop.session' }] }, 'electron', ['linux'])).not.toThrow();
        expect(() => validateNativeExtension({ ...valid, helper_providers: [{ owner: 'device-control', capability: 'arbitrary.exec' }] }, 'electron', ['linux'])).toThrowError(NativeExtensionCompatibilityError);
        expect(() => validateNativeExtension({ ...valid, helper_providers: [{ owner: 'device-control', capability: 'desktop.session', path: '/tmp/helper' }] }, 'electron', ['linux'])).toThrowError(NativeExtensionCompatibilityError);
    });
    it('renders the declaration as data', async () => {
        const generator = new DesktopTemplateGenerator({ framework: 'electron', output_path: '/tmp/native-generation-test', platforms: ['linux'], native_extension: valid, features: {}, window: {} } as unknown as DesktopConfig);
        const rendered = await (generator as any).processTemplateContent('JSON.parse("{{NATIVE_EXTENSION_CONFIG}}")', {});
        expect(JSON.parse(JSON.parse(rendered.slice('JSON.parse('.length, -1)))).toEqual(valid);
    });
});


describe('generated native extension layout', () => {
    it.each([false, true])('packages executable template hooks (enabled=%s)', async enabled => {
        const output = await fs.mkdtemp(path.join(os.tmpdir(), 'native-layout-'));
        try {
            const generator = new DesktopTemplateGenerator({ app_name: 'example', app_display_name: 'Example', app_description: 'Example',
                version: '1.0.0', author: 'test', license: 'MIT', app_id: 'test.example', framework: 'electron', template_type: 'basic',
                server_type: 'external', server_path: '', api_endpoint: '', output_path: output, platforms: ['linux'],
                features: {}, window: {}, update_config: { provider: 'none' },
                ...(enabled ? { native_extension: valid } : {}),
            } as DesktopConfig);
            const internal = generator as any;
            internal.templateBasePath = path.resolve(__dirname, '..');
            const template = await internal.loadTemplateConfig();
            await internal.processTemplateFiles(template);
            await internal.generateAdditionalFiles(template);
            const main = await fs.readFile(path.join(output, 'src/main.ts'), 'utf8');
            const preload = await fs.readFile(path.join(output, 'src/preload.ts'), 'utf8');
            expect(main).toContain('from "./native/presentation"');
            // PKG-08: remote scenario content runs in a sandboxed renderer and
            // cannot reach Node or an unrestricted host IPC surface.
            expect(main).toContain('nodeIntegration: false');
            expect(main).toContain('contextIsolation: true');
            expect(main).toContain('sandbox: true');
            expect(main).toContain('return { action: "deny" }');
            expect(main).not.toContain('{{NATIVE_EXTENSION_CONFIG}}');
            expect(preload).not.toContain('{{NATIVE_EXTENSION_CONFIG}}');
            expect((await fs.readFile(path.join(output, 'src/native/presentation.ts'), 'utf8'))).toContain('installPresentation');
            await expect(fs.stat(path.join(output, 'src/native/__tests__'))).rejects.toThrow();
            for (const source of [main, preload]) {
                const result = ts.transpileModule(source, { reportDiagnostics: true, compilerOptions: { target: ts.ScriptTarget.ES2020, module: ts.ModuleKind.CommonJS } });
                expect(result.diagnostics).toEqual([]);
            }
            if (enabled) {
                const metadata = JSON.parse(await fs.readFile(path.join(output, 'native-extension.json'), 'utf8'));
                expect(metadata.module).toBe('presentation');
                expect(metadata.main_entrypoint).toBe('dist/native/presentation.js');
                const pkg = JSON.parse(await fs.readFile(path.join(output, 'package.json'), 'utf8'));
                expect(pkg.build.files).toContain('native-extension.json');
                expect(metadata.native_dependencies).toEqual([]);
                expect(metadata.helper_providers).toEqual([]);
            } else {
                expect(preload).toContain('if (JSON.parse("null"))');
                await expect(fs.stat(path.join(output, 'native-extension.json'))).rejects.toThrow();
            }
        } finally { await fs.rm(output, { recursive: true, force: true }); }
    });
});

File: scenarios/scenario-to-desktop/api/generation/native_extension_test.go
package generation

import (
	"encoding/json"
	"errors"
	"testing"
)

func TestNativeExtensionAdmission(t *testing.T) {
	for _, tc := range []struct {
		name, raw string
		valid     bool
	}{
		{"vanilla", `{}`, true},
		{"presentation", `{"native_extension":{"version":1,"module":"presentation","permissions":["window.presentation"],"platforms":["linux"]}}`, true},
		{"future", `{"native_extension":{"version":3,"module":"presentation","permissions":["window.presentation"],"platforms":["linux"]}}`, false},
		{"import", `{"native_extension":{"version":1,"module":"presentation","entrypoint":"/tmp/code.js","permissions":["window.presentation"],"platforms":["linux"]}}`, false},
		{"permission", `{"native_extension":{"version":1,"module":"presentation","permissions":["filesystem"],"platforms":["linux"]}}`, false},
		{"target", `{"native_extension":{"version":1,"module":"presentation","permissions":["window.presentation"],"platforms":["mac"]}}`, false},
	} {
		t.Run(tc.name, func(t *testing.T) {
			config := DesktopConfig{Framework: "electron", Platforms: []string{"linux"}}
			err := json.Unmarshal([]byte(tc.raw), &config)
			if err == nil {
				err = config.ValidateNativeExtension()
			}
			if (err == nil) != tc.valid {
				t.Fatalf("valid=%v, err=%v", tc.valid, err)
			}
		})
	}
}

func TestNativeExtensionCompatibilityErrorIsTyped(t *testing.T) {
	config := DesktopConfig{
		Framework: "electron",
		Platforms: []string{"windows-arm64"},
		NativeExtension: &NativeExtension{
			Version:     1,
			Module:      "presentation",
			Permissions: []string{"window.presentation"},
			Platforms:   []string{"linux"},
		},
	}
	var compatibility *NativeExtensionCompatibilityError
	if err := config.ValidateNativeExtension(); !errors.As(err, &compatibility) {
		t.Fatalf("expected typed compatibility error, got %v", err)
	}
	if compatibility.Code != "NATIVE_EXTENSION_TARGET_UNSUPPORTED" || compatibility.Target != "windows-arm64" || compatibility.Field != "platforms" {
		t.Fatalf("unexpected compatibility error: %+v", compatibility)
	}
}

func TestNativeExtensionCanonicalPipelineTargets(t *testing.T) {
	config := DesktopConfig{Framework: "electron", NativeExtension: &NativeExtension{Version: 1, Module: "presentation", Permissions: []string{"window.presentation"}, Platforms: []string{"linux", "mac", "win"}}}
	for _, target := range []string{"linux-amd64", "linux-arm64", "darwin-amd64", "darwin-arm64", "windows-amd64", "windows-arm64"} {
		config.Platforms = []string{target}
		if err := config.ValidateNativeExtension(); err != nil {
			t.Fatalf("%s: %v", target, err)
		}
	}
	for _, target := range []string{"linux-unknown", "linux-amd64-extra", "unknown-amd64"} {
		config.Platforms = []string{target}
		if err := config.ValidateNativeExtension(); err == nil {
			t.Fatalf("accepted %s", target)
		}
	}
}

func TestNativeExtensionGovernedHelperProvider(t *testing.T) {
	config := DesktopConfig{Framework: "electron", Platforms: []string{"linux"}, NativeExtension: &NativeExtension{
		Version: 3, Module: "presentation", Permissions: []string{"window.presentation", "global-shortcut", "desktop.context"}, Platforms: []string{"linux"}, ActivationShortcut: "Control+Shift+Space",
		HelperProviders: []HelperProvider{{Owner: "device-control", Capability: "desktop.session"}},
	}}
	if err := config.ValidateNativeExtension(); err != nil {
		t.Fatalf("governed helper provider should be admitted: %v", err)
	}
	config.NativeExtension.HelperProviders[0].Capability = "arbitrary.exec"
	if err := config.ValidateNativeExtension(); err == nil {
		t.Fatal("arbitrary helper capability must be refused")
	}
}

func TestNativeActivationContract(t *testing.T) {
	config := DesktopConfig{Framework: "electron", Platforms: []string{"linux-amd64"}, NativeExtension: &NativeExtension{Version: 2, Module: "presentation", Permissions: []string{"window.presentation", "global-shortcut"}, Platforms: []string{"linux"}, ActivationShortcut: "CommandOrControl+Shift+Space"}}
	if err := config.ValidateNativeExtension(); err != nil {
		t.Fatal(err)
	}
	config.NativeExtension.Version = 1
	if err := config.ValidateNativeExtension(); err == nil {
		t.Fatal("v1 granted shortcut")
	}
	config.NativeExtension.Version = 2
	config.NativeExtension.ActivationShortcut = "Alt+Alt+Space"
	if err := config.ValidateNativeExtension(); err == nil {
		t.Fatal("duplicate modifiers accepted")
	}
}

func TestNativeContextPermission(t *testing.T) {
	c := DesktopConfig{Framework: "electron", Platforms: []string{"linux"}, NativeExtension: &NativeExtension{Version: 3, Module: "presentation", Permissions: []string{"window.presentation", "global-shortcut", "desktop.context"}, Platforms: []string{"linux"}, ActivationShortcut: "Control+Shift+Space"}}
	if err := c.ValidateNativeExtension(); err != nil {
		t.Fatal(err)
	}
	c.NativeExtension.Version = 2
	if c.ValidateNativeExtension() == nil {
		t.Fatal("older contract granted context")
	}
	c.NativeExtension.Version = 3
	c.NativeExtension.Permissions = []string{"window.presentation", "global-shortcut"}
	if c.ValidateNativeExtension() == nil {
		t.Fatal("missing context permission accepted")
	}
}
